### PR TITLE
Additional comments and testing related to SRTP resolution

### PR DESCRIPTION
**Pull request by dsyme**
In F# vNext I proposes we fix two oddities in resolving statically resolved method constraints in the presence of method overloading

See https://github.com/Microsoft/visualfsharp/issues/3814

The changes in test.fsx document the two cases a little further. It seems important to get these cases ironed out before https://github.com/Microsoft/visualfsharp/pull/3582 proceeds

I'm labeling this as F# vNext because it has an impact on type inference and SRTP method resolution in some corner cases, see the changes to the tests.  We will characterize these in an RFC.

### CODE REVIEW

**Comment by dsyme** [comment]
@gusty If you have a chance to run this over FSharp.Control or your most extreme uses of SRTP I'd be grateful

**Comment by gusty** [comment]
Thanks !
I faced this issue million times, am so happy you fixed it.
I will try to compile FSharpPlus including all tests and sample scripts plus some exotic snippets I have ot make sure it doesn't break anything else.
I'm 100% sure it will have an impact on that code base, but most likely it will be that it will simplify all the hackery I was forced to write in order to get the right overload resolution.
I will let you know of the outcome.

**Comment by dsyme** [comment]
@gusty I pushed a fix that you may need (indeed you may want to wait until this thing is green)

**Comment by gusty** [comment]
Yes, in any case I won't be able to try it today.
It also looks interesting all the doc you are including, it makes me understand better how all this works.

**Comment by dsyme** [comment]
Test failures: 
```

1) Failed : FSharp-Tests-Core+GeneratedSignatureTests.members-basics-GENERATED_SIGNATURE
Error running command 'D:\j\w\release_ci_pa---866fd2c3\tests\fsharp\..\..\release\net40\bin\fsc.exe' with args '-r:System.Core.dll --nowarn:20 --define:COMPILED --sig:tmptest.fsi --define:TESTS_AS_APP tmptest.fs' in directory 'D:\j\w\release_ci_pa---866fd2c3\tests\fsharp\core\members\basics'. ERRORLEVEL 1
at TestFramework.checkResult(CmdResult result) in D:\j\w\release_ci_pa---866fd2c3\tests\fsharp\test-framework.fs:line 258
at TestFramework.exec@438-22.Invoke(CmdResult result) in D:\j\w\release_ci_pa---866fd2c3\tests\fsharp\test-framework.fs:line 438
at TestFramework.fsc@446-1.Invoke(String flags, FSharpList`1 srcFiles) in D:\j\w\release_ci_pa---866fd2c3\tests\fsharp\test-framework.fs:line 446
at SingleTest.singleTestBuildAndRunCore(TestConfig cfg, String copyFiles, Permutation p) in D:\j\w\release_ci_pa---866fd2c3\tests\fsharp\single-test.fs:line 116
at FSharp-Tests-Core.GeneratedSignatureTests.members-basics-GENERATED_SIGNATURE() in D:\j\w\release_ci_pa---866fd2c3\tests\fsharp\tests.fs:line 2345

2) Error : FSharp-Tests-Core+TypecheckTests.type check neg99
System.Exception : neg99.err neg99.bsl differ; "diff between [D:\j\w\release_ci_pa---866fd2c3\tests\fsharp\typecheck\sigs\neg99.bsl] and [D:\j\w\release_ci_pa---866fd2c3\tests\fsharp\typecheck\sigs\neg99.err]
diff at line 4
- 
- neg99.fs(25,39,25,43): typecheck error FS0043: The type 'CrashFSC.OhOh.MyByte' does not support a conversion to the type 'CrashFSC.OhOh.MyByte'
"; neg99.vserr neg99.bsl differ; "diff between [D:\j\w\release_ci_pa---866fd2c3\tests\fsharp\typecheck\sigs\neg99.bsl] and [D:\j\w\release_ci_pa---866fd2c3\tests\fsharp\typecheck\sigs\neg99.vserr]
diff at line 4
- 
- neg99.fs(25,39,25,43): typecheck error FS0043: The type 'CrashFSC.OhOh.MyByte' does not support a conversion to the type 'CrashFSC.OhOh.MyByte'
"
   at Microsoft.FSharp.Core.PrintfModule.PrintFormatToStringThenFail@1645.Invoke(String message) in D:\j\w\release_ci_pa---866fd2c3\src\fsharp\FSharp.Core\printf.fs:line 1645

3) Failed : FSharp-Tests-Core+CoreTests.members-basics-AS_DLL
Error running command 'D:\j\w\release_ci_pa---866fd2c3\tests\fsharp\..\..\release\net40\bin\fsc.exe' with args '-r:System.Core.dll --nowarn:20 --define:COMPILED --optimize -a -o:test--optimize-lib.dll -g test.fsi test.fs' in directory 'D:\j\w\release_ci_pa---866fd2c3\tests\fsharp\core\members\basics'. ERRORLEVEL 1
at TestFramework.checkResult(CmdResult result) in D:\j\w\release_ci_pa---866fd2c3\tests\fsharp\test-framework.fs:line 258
at TestFramework.exec@438-22.Invoke(CmdResult result) in D:\j\w\release_ci_pa---866fd2c3\tests\fsharp\test-framework.fs:line 438
at TestFramework.fsc@446-1.Invoke(String flags, FSharpList`1 srcFiles) in D:\j\w\release_ci_pa---866fd2c3\tests\fsharp\test-framework.fs:line 446
at SingleTest.singleTestBuildAndRunCore(TestConfig cfg, String copyFiles, Permutation p) in D:\j\w\release_ci_pa---866fd2c3\tests\fsharp\single-test.fs:line 154
at FSharp-Tests-Core.CoreTests.members-basics-AS_DLL() in D:\j\w\release_ci_pa---866fd2c3\tests\fsharp\tests.fs:line 991

4) Failed : FSharp-Tests-Core+CoreTests.members-basics-FSC_BASIC
Error running command 'D:\j\w\release_ci_pa---866fd2c3\tests\fsharp\..\..\release\net40\bin\fsc.exe' with args '-r:System.Core.dll --nowarn:20 --define:COMPILED --optimize+ --debug -o:test--optplus--debug.exe -g test.fsi test.fs' in directory 'D:\j\w\release_ci_pa---866fd2c3\tests\fsharp\core\members\basics'. ERRORLEVEL 1
at TestFramework.checkResult(CmdResult result) in D:\j\w\release_ci_pa---866fd2c3\tests\fsharp\test-framework.fs:line 258
at TestFramework.exec@438-22.Invoke(CmdResult result) in D:\j\w\release_ci_pa---866fd2c3\tests\fsharp\test-framework.fs:line 438
at TestFramework.fsc@446-1.Invoke(String flags, FSharpList`1 srcFiles) in D:\j\w\release_ci_pa---866fd2c3\tests\fsharp\test-framework.fs:line 446
at SingleTest.singleTestBuildAndRunCore(TestConfig cfg, String copyFiles, Permutation p) in D:\j\w\release_ci_pa---866fd2c3\tests\fsharp\single-test.fs:line 140
at FSharp-Tests-Core.CoreTests.members-basics-FSC_BASIC() in D:\j\w\release_ci_pa---866fd2c3\tests\fsharp\tests.fs:line 987

5) Failed : FSharp-Tests-Core+CoreTests.members-basics-FSI_BASIC
Error running command 'D:\j\w\release_ci_pa---866fd2c3\tests\fsharp\..\..\release\net40\bin\fsi.exe' with args '-r:System.Core.dll --nowarn:20 --define:INTERACTIVE --maxerrors:1 --abortonerror test.fsi test.fs' in directory 'D:\j\w\release_ci_pa---866fd2c3\tests\fsharp\core\members\basics'. ERRORLEVEL 1
at TestFramework.checkResult(CmdResult result) in D:\j\w\release_ci_pa---866fd2c3\tests\fsharp\test-framework.fs:line 258
at TestFramework.exec@438-22.Invoke(CmdResult result) in D:\j\w\release_ci_pa---866fd2c3\tests\fsharp\test-framework.fs:line 438
at TestFramework.fsi@457-3.Invoke(String flags, FSharpList`1 sources) in D:\j\w\release_ci_pa---866fd2c3\tests\fsharp\test-framework.fs:line 457
at SingleTest.singleTestBuildAndRunCore(TestConfig cfg, String copyFiles, Permutation p) in D:\j\w\release_ci_pa---866fd2c3\tests\fsharp\single-test.fs:line 95
at FSharp-Tests-Core.CoreTests.members-basics-FSI_BASIC() in D:\j\w\release_ci_pa---866fd2c3\tests\fsharp\tests.fs:line 983

6) Failed : FSharp-Tests-Core+CoreTests.members-basics-hw
Error running command 'D:\j\w\release_ci_pa---866fd2c3\tests\fsharp\..\..\release\net40\bin\fsc.exe' with args '-r:System.Core.dll --nowarn:20 --define:COMPILED --optimize+ --debug -o:test--optplus--debug.exe -g test.fsx' in directory 'D:\j\w\release_ci_pa---866fd2c3\tests\fsharp\core\members\basics-hw'. ERRORLEVEL 1
at TestFramework.checkResult(CmdResult result) in D:\j\w\release_ci_pa---866fd2c3\tests\fsharp\test-framework.fs:line 258
at TestFramework.exec@438-22.Invoke(CmdResult result) in D:\j\w\release_ci_pa---866fd2c3\tests\fsharp\test-framework.fs:line 438
at TestFramework.fsc@446-1.Invoke(String flags, FSharpList`1 srcFiles) in D:\j\w\release_ci_pa---866fd2c3\tests\fsharp\test-framework.fs:line 446
at SingleTest.singleTestBuildAndRunCore(TestConfig cfg, String copyFiles, Permutation p) in D:\j\w\release_ci_pa---866fd2c3\tests\fsharp\single-test.fs:line 140
at FSharp-Tests-Core.CoreTests.members-basics-hw() in D:\j\w\release_ci_pa---866fd2c3\tests\fsharp\tests.fs:line 995

7) Failed : FSharp-Tests-Core+CoreTests.members-basics-hw-mutrec
Error running command 'D:\j\w\release_ci_pa---866fd2c3\tests\fsharp\..\..\release\net40\bin\fsc.exe' with args '-r:System.Core.dll --nowarn:20 --define:COMPILED --optimize+ --debug -o:test--optplus--debug.exe -g test.fs' in directory 'D:\j\w\release_ci_pa---866fd2c3\tests\fsharp\core\members\basics-hw-mutrec'. ERRORLEVEL 1
at TestFramework.checkResult(CmdResult result) in D:\j\w\release_ci_pa---866fd2c3\tests\fsharp\test-framework.fs:line 258
at TestFramework.exec@438-22.Invoke(CmdResult result) in D:\j\w\release_ci_pa---866fd2c3\tests\fsharp\test-framework.fs:line 438
at TestFramework.fsc@446-1.Invoke(String flags, FSharpList`1 srcFiles) in D:\j\w\release_ci_pa---866fd2c3\tests\fsharp\test-framework.fs:line 446
at SingleTest.singleTestBuildAndRunCore(TestConfig cfg, String copyFiles, Permutation p) in D:\j\w\release_ci_pa---866fd2c3\tests\fsharp\single-test.fs:line 140
at FSharp-Tests-Core.CoreTests.members-basics-hw-mutrec() in D:\j\w\release_ci_pa---866fd2c3\tests\fsharp\tests.fs:line 999

8) Failed : FSharp-Tests-Core+CoreTests.members-ops
Error running command 'D:\j\w\release_ci_pa---866fd2c3\tests\fsharp\..\..\release\net40\bin\fsc.exe' with args '-r:System.Core.dll --nowarn:20 --define:COMPILED --optimize+ --debug -o:test--optplus--debug.exe -g test.fsx' in directory 'D:\j\w\release_ci_pa---866fd2c3\tests\fsharp\core\members\ops'. ERRORLEVEL 1
at TestFramework.checkResult(CmdResult result) in D:\j\w\release_ci_pa---866fd2c3\tests\fsharp\test-framework.fs:line 258
at TestFramework.exec@438-22.Invoke(CmdResult result) in D:\j\w\release_ci_pa---866fd2c3\tests\fsharp\test-framework.fs:line 438
at TestFramework.fsc@446-1.Invoke(String flags, FSharpList`1 srcFiles) in D:\j\w\release_ci_pa---866fd2c3\tests\fsharp\test-framework.fs:line 446
at SingleTest.singleTestBuildAndRunCore(TestConfig cfg, String copyFiles, Permutation p) in D:\j\w\release_ci_pa---866fd2c3\tests\fsharp\single-test.fs:line 140
at FSharp-Tests-Core.CoreTests.members-ops() in D:\j\w\release_ci_pa---866fd2c3\tests\fsharp\tests.fs:line 118

9) Failed : FSharp-Tests-Core+CoreTests.members-ops-mutrec
Error running command 'D:\j\w\release_ci_pa---866fd2c3\tests\fsharp\..\..\release\net40\bin\fsc.exe' with args '-r:System.Core.dll --nowarn:20 --define:COMPILED --optimize+ --debug -o:test--optplus--debug.exe -g test.fs' in directory 'D:\j\w\release_ci_pa---866fd2c3\tests\fsharp\core\members\ops-mutrec'. ERRORLEVEL 1
at TestFramework.checkResult(CmdResult result) in D:\j\w\release_ci_pa---866fd2c3\tests\fsharp\test-framework.fs:line 258
at TestFramework.exec@438-22.Invoke(CmdResult result) in D:\j\w\release_ci_pa---866fd2c3\tests\fsharp\test-framework.fs:line 438
at TestFramework.fsc@446-1.Invoke(String flags, FSharpList`1 srcFiles) in D:\j\w\release_ci_pa---866fd2c3\tests\fsharp\test-framework.fs:line 446
at SingleTest.singleTestBuildAndRunCore(TestConfig cfg, String copyFiles, Permutation p) in D:\j\w\release_ci_pa---866fd2c3\tests\fsharp\single-test.fs:line 140
at FSharp-Tests-Core.CoreTests.members-ops-mutrec() in D:\j\w\release_ci_pa---866fd2c3\tests\fsharp\tests.fs:line 121
```

**Comment by dsyme** [comment]
Adding the WIP label as this is going to require care to make progress on - there are cases where the existing proposed fix causes a change in type inference behavior which we need to asses, possibly via an RFC

**Comment by gusty** [comment]
@dsyme  I tried compiling [FSharpPlus](https://github.com/gusty/FSharpPlus) with this fix-3814 and it fails, specifically compiling [this line](https://github.com/gusty/FSharpPlus/blob/c611e4777f039865cf6c17855c6a1f295db51ce7/src/FSharpPlus/Traversable.fs#L75) which seems to be a return-type-directed one. Here's the error message:

    C:\Repos\FSharpPlus\src\FSharpPlus\Traversable.fs(75,137,75,167): typecheck error FS0332: Could not resolve the ambiguity inherent in the use of the operator 'Map' at or near this program point. Consider using type annotations to resolve the ambiguity.

I also tried compiling some scripts using the [CI00091 release](https://www.nuget.org/packages/FSharpPlus/1.0.0-CI00091):

 - [Converter.fsx](https://github.com/gusty/FSharpPlus/blob/c611e4777f039865cf6c17855c6a1f295db51ce7/src/FSharpPlus/SamplesFromFsControl/Converter.fsx#L103) fails on that line (just the last line) with:

```
Converter.fsx(103,48): error FS0072: Lookup on object of indeterminate type based on information prior to this program point. A type annotation may be needed prior to this program point to constrain the type of the object. This may allow the lookup to be resolved.
```
 - [Functions.fsx](https://github.com/gusty/FSharpPlus/blob/c611e4777f039865cf6c17855c6a1f295db51ce7/src/FSharpPlus/SamplesFromFsControl/Functions.fsx#L15) fails on a simple ``map`` with:

```
Functions.fsx(15,21): error FS0332: Could not resolve the ambiguity inherent in the use of the operator 'Map' at or near this program point. Consider using type annotations to resolve the ambiguity.

Functions.fsx(15,21): error FS0071: Type constraint mismatch when applying the default type 'obj' for a type inference variable. This expression was expected to have type
    'obj'
but here has type
    'int option'     Consider adding further type constraints
```

 - [Collections.fsx](https://github.com/gusty/FSharpPlus/blob/c611e4777f039865cf6c17855c6a1f295db51ce7/src/FSharpPlus/SamplesFromFsControl/Collections.fsx#L54) fails with:

```
Collections.fsx(54,70): error FS0332: Could not resolve the ambiguity inherent in the use of the operator 'Plus' at or near this program point. Consider using type annotations to resolve the ambiguity.

Collections.fsx(54,70): error FS0071: Type constraint mismatch when applying the default type 'obj' for a type inference variable. No overloads match for method 'Plus'. The available overloads are shown below. Consider adding further type constraints
Possible overload: 'static member Plus.Plus : x: ^Plus * y: ^Plus * Internals.Default2 ->  ^Plus when  ^Plus : (static member Append :  ^Plus *  ^Plus ->  ^Plus)'. Type constraint mismatch. The type
    'obj'
is not compatible with type
    ''a'
.
Possible overload: 'static member Plus.Plus : x: ^Plus * y: ^Plus * _mthd:Internals.Default1 ->  ^Plus when  ^Plus : (static member ( + ) :  ^Plus *  ^Plus ->  ^Plus)'. Type constraint mismatch. The type
    'obj'
is not compatible with type
    ''a'
.
Possible overload: 'static member Plus.Plus :  ^t *  ^t * _mthd:Internals.Default1 -> ('a1 -> 'a1) when  ^t : null and  ^t : struct'. Type constraint mismatch. The type
    'obj'
is not compatible with type
    ''a'
.
Possible overload: 'static member Plus.Plus : x:'a0 list * y:'a0 list * _mthd:Plus -> 'a0 list'. Type constraint mismatch. The type
    'obj'
is not compatible with type
    ''a list'
.
Possible overload: 'static member Plus.Plus : x:'a0 array * y:'a0 [] * _mthd:Plus -> 'a0 []'. Type constraint mismatch. The type
    'obj'
is not compatible with type
    ''a array'
.
Possible overload: 'static member Plus.Plus : unit * unit * _mthd:Plus -> unit'. Type constraint mismatch. The type
    'obj'
is not compatible with type
    'unit'
...
```

Then I stopped testing, since most error messages seems to be more or less the same. Trying to match default inferred ``obj`` with something concrete and failing.

So this seems to be an important breaking change in type inference but maybe there's a way to make it backward compatible and avoid the RFC and the breaking change at all.

**Comment by gusty** [comment]
I came up with a minimal repro:

    type Map =
        static member inline Invoke (mapping :'T->'U) (source : '``Functor<'T>``) : '``Functor<'U>`` = 
            let inline call (mthd : ^M, source : ^I, _output : ^R) = ((^M or ^I or ^R) : (static member Map: _*_*_ -> _) source, mapping, mthd)
            call (Unchecked.defaultof<Map>, source, Unchecked.defaultof<'``Functor<'U>``>)

        static member Map (x : option<_>      , f : 'T->'U, _mthd : Map) = Option.map  f x
        static member Map (x : list<_>        , f : 'T->'U, _mthd : Map) = List.map f x

    let inline map f x = Map.Invoke f x
   
    // test
    map string [1]

Note that if you remove completely ``^R`` still fails, this time with a Value Restriction.

**Comment by realvictorprm** [comment]
@dsyme is there already an RFC open for this or can we make one as soon as possible? According the office hours this is a crucial point which needs to be fixed before other SRTP features can be merged at all.

**Comment by realvictorprm** [comment]
Ok @dsyme I'm looking into this right now. I hope to be able to fix this soon, maybe we're lucky ¯\_(ツ)_/¯

**Comment by realvictorprm** [comment]
@dsyme OK. I want to give this a try again however last time I failed to proceed much because I'm not understanding really what's happening really in the overload resolution there. Would it possible to provide here an even more detailed description of why we have a weak and strong resolution in the methods processing the traits?

**Comment by realvictorprm** [comment]
Hey @dsyme to solve the first case of the problem your inital code was doing the right job. Moreover I went through it again (after learning more about type inference and the F# compiler) and the changes do make sense to me in the particular failing case. It's true that no resolution of members on a not more specified type should happen or at least it shouldn't give back the object methods. Can we discuss this change/fix again?

**Comment by dsyme** [comment]
From slack

> @dsyme says  IIRC the problem is that the fix still caused changes in inference behaviour which were beyond a bug fix.  I'll take another look today

**Comment by dsyme** [comment]
@realvictorprm and I have looked at this and decided it contained three things

1. code cleanup, comments and renaming related to SRTP
2. additional testing related to both the issues being fixed and pinning down some existing behaviour
3. the fix itself 

We decided to get parts (1) and (2) in the bank before trying (3), so this is reverted to be a no-op PR containing only code cleanup, comments and additional testing.

**Comment by dsyme** [comment]
This is now ready to merge - as described above it is extra "pin down current behaviour" testing and comments

**Comment by dsyme** [comment]
> Approval for F# 5.0 - though we'll obviously flight this in nightlies long before a preview of a release gets cut

There's no change in behaviour, it's just comment and testing, it can go straight into master.

File: tests/fsharp/core/subtype/test.fsx [hunk 1]
```diff
@@ -232,8 +232,31 @@ module SomeRandomOperatorConstraints = begin
 
     let f2 x : float = x * x 
     let f3 x (y:float) = x * y
+
     //let neg4 x (y:System.DateTime) = x + y
+
+    // This example resolves the type of "y" to "TimeSpam". It checks that a single "+" overload between 
+    // two different types DateTime andd TimeSpan gets resolved via
```
Comment (gusty) [inline tests/fsharp/core/subtype/test.fsx]: Type here

File: tests/fsharp/core/subtype/test.fsx [hunk 2]
```diff
@@ -232,8 +232,31 @@ module SomeRandomOperatorConstraints = begin
 
     let f2 x : float = x * x 
     let f3 x (y:float) = x * y
+
     //let neg4 x (y:System.DateTime) = x + y
+
+    // This example resolves the type of "y" to "TimeSpam". It checks that a single "+" overload between 
+    // two different types DateTime andd TimeSpan gets resolved via 
+    // vis weak SRTP resolution using a DateTime constraint alone.
```
Comment (gusty) [inline tests/fsharp/core/subtype/test.fsx]: ... and here

File: tests/fsharp/core/subtype/test.fsx [hunk 3]
```diff
@@ -232,8 +232,31 @@ module SomeRandomOperatorConstraints = begin
...
     let f5 (x:DateTime) y = x + y
+
+    // This example checks a use of TimeSpan/DateTime overloads
+    let f5b (x:DateTime) (y:DateTime) = (x - y) 
+
+
+    // This example checks a use of TimeSpan/DateTime overloads
+    let f5b2 (x:DateTime) (y:TimeSpan) = (x - y) 
+
+    // This example coincidentally checks that the return type is not taken into account before th list of method overloads
```
Comment (gusty) [inline tests/fsharp/core/subtype/test.fsx]: Also here.
